New recipe for pgmacs.

emarsden opened the pull request:
### Brief summary of what the package does

This package provides a browsing and editing interface for the PostgreSQL database.

### Direct link to the package repository

https://github.com/emarsden/pgmacs

### Your association with the package

Maintainer.

### Relevant communications with the upstream package maintainer

That's me.

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] I've used `M-x checkdoc` to check the package's documentation strings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

This package is in a fairly early state, but I believe it would be more convenient for users to be able to install it from MELPA. There is an issue to be resolved concerning use of the vtable library, where we need to override a function to get recent behaviour. This is currently done using mode-specific advice. Perhaps a fork of vtable might be preferable, but I haven't yet moved in that direction.

riscy commented:
Thanks for this.  I acknowledge this package is in an early state.  Here's a quick first pass --

`pgmacs.el` with byte-compile using Emacs 29.2:
```
pgmacs.el:50:2: Warning: custom-declare-variable `pgmacs-row-limit' docstring wider than 80 characters
In pgmacs--insert-row/widget:
pgmacs.el:457:2: Warning: docstring wider than 80 characters
In pgmacs--yank-row:
pgmacs.el:534:2: Warning: docstring wider than 80 characters
In pgmacs--display-table:
pgmacs.el:787:2: Warning: docstring wider than 80 characters
In end of data:
pgmacs.el:391:10: Warning: the function `widget-forward' is not known to be defined.
```

`pgmacs.el` with checkdoc 0.6.2 (fix *within reason*):
```
pgmacs.el:51: Some lines are over 80 columns wide
pgmacs.el:102: All variables and subroutines might as well have a documentation string
pgmacs.el:107: All variables and subroutines might as well have a documentation string
pgmacs.el:111: All variables and subroutines might as well have a documentation string
pgmacs.el:241: All variables and subroutines might as well have a documentation string
pgmacs.el:686: All variables and subroutines might as well have a documentation string
pgmacs.el:1288: All variables and subroutines might as well have a documentation string
pgmacs.el:1333: All variables and subroutines might as well have a documentation string
pgmacs.el:716: `y-or-n-p' argument should end with "?"
pgmacs.el:774: White space found at end of line
```

`pgmacs.el` with package-lint 20240330.1458:
```
6 issues found:
240:0: error: `pgmacs--read-value/minibuffer' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
254:0: error: `pgmacs--edit-value/minibuffer' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
322:0: error: `pgmacs--edit-value/widget' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
457:0: error: `pgmacs--insert-row/widget' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
1197:0: error: `pgmacs-open/string' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
1209:0: error: `pgmacs-open/uri' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
```

`pgmacs.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- pgmacs.el#L432: This double negation can be collapsed (`not` aliases `null`)
- pgmacs.el#L474: This double negation can be collapsed (`not` aliases `null`)
- pgmacs.el#L555: This double negation can be collapsed (`not` aliases `null`)
- pgmacs.el#L653: Use `when ...` instead of `unless (null ...)`
- pgmacs.el#L1337: Loading a package should not add advice
- pgmacs.el#L140: Consider using %F instead of %Y-%m-%d in time strings
```

emarsden commented:
Thanks. I've fixed all these issues, other than the non-standard separator '/' in an identifier, which I prefer to retain.

riscy commented:
Thanks for accommodating.  The community style is to avoid `/` and other various nonstandard separators, and there's a reason it's explicitly checked for.

I notice this recipe is now also bringing in xtable.el which is outside of pgmacs's namespace; we should `:exclude` it or, alternatively, move that code into its own repository.

emarsden commented:
I have now changed the API to follow this conservative elisp style guide. 

I have renamed the xtable.el to pgmacstbl.el to include it in the package namespace, and added a comment indicating that this is intended as a purely private API.

riscy commented:
Thanks for accommodating - know it's appreciated.  Here's the latest salient issues since the changes --

`pgmacs.el` with byte-compile using Emacs 29.3:
```
pgmacs.el:1199:37: Warning: the function `pgmacstbl-column' is not known to be defined.
pgmacs.el:961:8: Warning: the function `pgmacstbl-goto-column' is not known to be defined.
pgmacs.el:959:8: Warning: the function `pgmacstbl-goto-object' is not known to be defined.
pgmacs.el:948:19: Warning: the function `text-property-search-backward' is not known to be defined.
pgmacs.el:934:10: Warning: the function `pgmacstbl-insert' is not known to be defined.
pgmacs.el:849:24: Warning: the function `make-pgmacstbl' is not known to be defined.
pgmacs.el:840:31: Warning: the function `make-pgmacstbl-column' is not known to be defined.
pgmacs.el:703:27: Warning: the function `pgmacstbl--limit-string' is not known to be defined.
pgmacs.el:305:10: Warning: the function `pgmacstbl-remove-object' is not known to be defined.
pgmacs.el:304:10: Warning: the function `pgmacstbl-insert-object' is not known to be defined.
pgmacs.el:287:31: Warning: the function `pgmacstbl-column-formatter' is not known to be defined.
pgmacs.el:275:23: Warning: the function `pgmacstbl-current-column' is not known to be defined.
pgmacs.el:272:28: Warning: the function `pgmacstbl-current-object' is not known to be defined.
pgmacs.el:243:19: Warning: the function `pgmacstbl-column-name' is not known to be defined.
pgmacs.el:238:17: Warning: the function `pgmacstbl-columns' is not known to be defined.
pgmacs.el:237:22: Warning: the function `pgmacstbl-current-table' is not known to be defined.
```

`pgmacstbl.el` with package-lint 20240330.1458 and `package-lint-main-file` = "pgmacs.el":
```
3 issues found:
1:0: error: There is no `provide' form.
5:0: error: Package-Requires outside the main file have no effect.
86:0: error: "make-pgmacstbl" doesn't start with package's prefix "pgmacs".
```

`pgmacstbl.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- pgmacstbl.el#L215: Triple semicolons `;;;` are usually for section headings
- pgmacstbl.el#L3: Have you done the paperwork to assign this copyright?
```

Package and license:
- pgmacstbl.el -- not in package namespace `pgmacs-`

And note there is a license mismatch in these files' boilerplate:
- GNU General Public License v2.0 or later -- pgmacs.el: Emacs is editing a PostgreSQL database
- GNU General Public License v3.0 or later -- pgmacstbl.el: Displaying data in tables

riscy commented:
Re: the note on the copyright, if you're just crediting code you're reusing then that's fine, and you can disregard my note.

emarsden commented:
The undefined function warnings were due to my misunderstanding how require/provide should be used between files in a same package. I think I've fixed that now, thanks.

The copyright issue is as it should be: pgmacstbl.el is almost all code reused from vtable.el in GNU Emacs, written by larsi. This is noted in the file header.

The vtable.el code has the same points (use of make-foo function and ;;; comments) as package-lint and melpazoid are complaining about. I'm hoping that if it's acceptable for Emacs core, it's acceptable for MELPA.

riscy commented:
> I'm hoping that if it's acceptable for Emacs core, it's acceptable for MELPA

This is sometimes not the case - and *also* what was once acceptable for MELPA isn't necessarily acceptable today.  But I try to err on the side of the official Emacs Lisp style guide.

I am not a lawyer but I *feel* like you don't have a shaky license setup here.  In summary:
- You are repackaging and dependent on GPLv3-or-later code as part of this program, i.e. pgmacstbl.el
- You are releasing the end-result, the `pgmacs` program, as GPLv2-or-later

I think you need to release `pgmacs` under GPLv3-or-later.

This is minor, but you also write: `For this reason, please don't use this code from outside of PGmacs.` but I might say `For this reason, it is better to depend on the builtin vtable than on this code.`  Not a blocker, but I bring this up because these licenses are pretty explicit about code reuse.

Also, as earlier, pgmacstbl.el is not in the package namespace `pgmacs-`; you should rename this file to `pgmacs-tbl.el` or `pgmacs-vtable.el` to make the link to vtable.el clear.

riscy commented:
Apologies, I missed reminding about this package-lint too --

`pgmacstbl.el` with package-lint 20240330.1458 and `package-lint-main-file` = "pgmacs.el":
```
1 issue found:
85:0: error: "make-pgmacstbl" doesn't start with package's prefix "pgmacs".
```

It may feel inconvenient, but a simple and fair way to prevent namespace collisions across MELPA (and hopefully elsewhere) is to ask that the package, variable names, and function names all use the same prefix.  In very rare cases I would make exceptions to this rule (e.g. backward compatibility with a package predating this rule) -- and `package-lint` has some of these exceptions built into it.
